nucypher/acumen/perception.py

```diff
@@ -168,8 +218,91 @@ def abridged_state_details(state):
+        if layer is None:
+            existing_strategies.append(strategy)
+        else:
+            existing_strategies.insert(strategy, index=layer)
+
+    def get_nodes(self, label=None) -> Iterator["Teacher"]:
+        """If label is None return all know nodes"""
+        if not label:
+            return iter(self)
+        if label not in self.BUCKETS:
+            raise self.UnknownLabel(f'{label} is not a valid node label')
+        try:
+            nodes = self.__marked[label]
+        except KeyError:
+            return iter(dict())  # empty
+        return iter(nodes)
+
+    def label(self, label, node: "Teacher") -> None:
+        """
+        Apply a label to a known node.  Removes any exiting labels before adding the new one.
+        If the provided label is not valid UnknownLabel is raised.
+        If the provided node is not known UnknownNode is raised.
+        """
+        if label not in self.BUCKETS:
+            raise self.UnknownLabel(f'{label} is not a valid node label')
+        if self.__nodes.get(node.checksum_address):
+            self.unlabel(node=node)            # Remove any existing labels...
+            self.__marked[label].append(node)  # Set the new label
+        else:
+            raise self.UnknownNode(f'Cannot label an unknown node ({node}).')
+
+    def unlabel(self, node: "Teacher", label=None) -> None:
+        """Removes a label from a node, or if label is None, all labels are removed."""
+
+        # Remove one label
+        if label:
+            if node in self.__marked[label]:
+                self.__marked[label].remove(node)
+                return
+
+        # Remove all labels
+        for _label in self.BUCKETS:
+            if node in self.__marked[_label]:
+                self.__marked[_label].remove(node)
+
+    def prune_bucket(self, label):
+        """Apply pruning strategies to a single node bucket once"""
+        try:
+            strategies = self.__pruning_strategies[label]
+        except KeyError:
+            raise self.UnknownLabel
+        for node in self.get_nodes(label=label):
+            for strategy in strategies:
+                keep = strategy(node=node)
+                if not keep:
+                    del self.__nodes[node.checksum_address]  # prune node
+                    self.unlabel(node=node, label=label)     # prune corresponding label
+                    break  # this node is already doomed anyways
+                    # TODO: forget node from disk too
+                    # TODO: Trash can label?
+            else:
+                # Reinstate this node to good standing by un/relabeling
+                if label in self.__pruning_strategies:
+                    self.unlabel(node=node, label=label)
+                    self.__reset_node_label_tracking(node=node)
+
+    def prune_nodes(self) -> None:
+        """Apply node pruning strategies to all marked nodes once"""
+        self._pruning_strategies: Dict[type, List[Callable]]
```

Comment:
Was it supposed to be `__pruning_strategies`?

Reply:
Yes, you are right - Ill include this in my next commit ( unless you beat me to it :wink:  )

Reply:
:heavy_check_mark: